lib/pc_patch.c

```diff
@@ -660,3 +660,66 @@ pc_patch_set_schema(const PCPATCH *patch, const PCSCHEMA *new_schema)
 
     return (PCPATCH*) paout;
 }
+
+/**
+* Rotate a patch given a unit quaternion.
+*/
+PCPATCH *
+pc_patch_rotate_quaternion(
+    const PCPATCH *patch,
+    double qw, double qx, double qy, double qz,
+    const char *xdimname, const char *ydimname, const char *zdimname)
+{
+    PCPATCH_UNCOMPRESSED *uncompressed_patch;
+    PCDIMENSION *xdim, *ydim, *zdim;
+    const PCSCHEMA *schema;
+    PCPOINTLIST *pointlist;
+    PCMAT33 qmat;
+    PCVEC3 vec, rvec;
+    size_t i;
+
+    pc_matrix_set_from_quaternion(qmat, qw, qx, qy, qz);
+
+    if ( patch->type == PC_NONE )
+    {
```

Comment:
Concerning the api, that pattern is already used in the lib, for instance :
[Conditionnal clone](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L578)
[Conditionnal free](https://github.com/pgpointcloud/pointcloud/blob/3929653f51296f5dd5fe5997c8c4c5d46419cb50/pgsql/pc_access.c#L622)

a `pc_patch_uncompressed_clone` function (and its siblings) would enable more readable code and optimized execution.